### PR TITLE
Upgrade USWDS

### PR DESCRIPTION
**Pull request by ncksllvn**
This contains all of the changes necessary to upgrade USWDS.

### TODO
- [x] Fix buttons
    - The `usa-button-outline` class has been renamed to `usa-button-secondary`
- [x] Fix grid
    - Not sure what's going on with the sidebar pages
- [x] Radio boxes
    - For some reason Nightwatch can no longer click radio boxes
- [x] The CSS helper for `[hidden]`  now uses the `!important` directive, so it's tougher to override by any widgets that use it. Specifically, the education widget uses this and can't override it using the `data-state="open"` anymore.

#### TODOs continued
All items that should be fixed as a result of this upgrade should be noted here:

department-of-veterans-affairs/vets.gov-team/issues/5622

#### Foundation vs. USWDS naming conflict
There was a variable naming conflict between USWDS and Foundation, these variables being `$small`, `$medium`, and `$large`. It was not fun finding these, but it's kinda funny that they were defined in Foundation's code only for legacy purposes, and were new in USWDS. I tracked them down from this error:

```
@media screen and ($default-feature: nth($query, 1)) {
                                                 ^
      (min-width: 64.0625em) isn't a valid CSS value.
```
This code was part of the `media` mixin. This meant that at some point it was being called with a value express in `em`'s. After trying some other things I started digging through Foundation's variable declarations to find anything that might raise an alarm and that's where I found those variables that match USWDS's. We include USWDS's sass after Foundation's, but USWDS's declares theirs using the `!default` modifier, meaning that the declaration only takes effect if they aren't already declared at that point, so Foundation's were still taking precedence. So, after including Foundation in our `style.scss`, I also added USWDS's declarations for `$small`, `$medium`, and `$large` without that `!default` modifier.

#### Bourbon and Neat problem
There was also another small problem with `bourbon` and `bourbon-neat` being required in our project without being in our `package.json`. We were able to directly require them in our JS with the old USWDS version but we can't anymore, so I commented them out of our `webpack.config.js`.

##### Main Issue
Resolves department-of-veterans-affairs/vets.gov-team/issues/5591

##### Bugs
Resolves department-of-veterans-affairs/vets.gov-team/issues/5622

##### Other Issues
Resolves department-of-veterans-affairs/vets.gov-team/issues/5591
Resolves department-of-veterans-affairs/vets.gov-team/issues/2873
Resolves department-of-veterans-affairs/vets.gov-team/issues/1463

### Intentional diversions from USWDS
If there are deliberate design decisions to differ from what USWDS offers and you want to make note of that, you can document them here - department-of-veterans-affairs/vets.gov-team/issues/5618

### CODE REVIEW

File: src/sass/style.scss [hunk 1]
```diff
@@ -28,7 +28,6 @@
 
 @import "~uswds/src/stylesheets/core/defaults";
 @import "~uswds/src/stylesheets/core/base";
-@import "~uswds/src/stylesheets/core/grid-settings";
```
Comment (ncksllvn) [inline src/sass/style.scss]: This import is no longer used in USWDS.

**Review by cehsu — COMMENTED** [review]
Thanks for your research and work on this, and for the detailed explanation, I asked for some additional insights on a few things.

File: config/webpack.config.js [hunk 2]
```diff
@@ -4,8 +4,8 @@ const ExtractTextPlugin = require('extract-text-webpack-plugin');
 const ManifestPlugin = require('webpack-manifest-plugin');
 const ChunkManifestPlugin = require('chunk-manifest-webpack-plugin');
 const WebpackMd5Hash = require('webpack-md5-hash');
-const bourbon = require('bourbon').includePaths;
-const neat = require('bourbon-neat').includePaths;
+// const bourbon = require('bourbon').includePaths;
+// const neat = require('bourbon-neat').includePaths;
```
[the remark this answers is not in the document]
Comment (ncksllvn) [reply config/webpack.config.js]: Yeah I believe that it resolves the same problem you have linked. With the previous version of USWDS, it would install bourbon and neat in the top-level of our node_modules folder, so we could require it this way. But with the upgrade, it isn't top-level anymore. Not exactly sure how that works, but I'm finding that to be the case. 

Our SASS style includes Bourbon and Neat from USWDS's lib directory (see the top of style.scss) so I don't think that costs us any anything - we still have the grid and all of USWDS style. I'm not exactly sure what these do in our Webpack but maybe @jbalboni knows for sure. 

If we find these to be important for one reason or another we can just add bourbon and neat to our `package.json` and uncomment those lines. I didn't do that because USWDS is using such specific [versions](https://github.com/18F/web-design-standards/blob/develop/package.json#L74).

File: src/sass/base/_b-breakpoints.scss [hunk 3]
```diff
@@ -1,9 +1,13 @@
...
 // Additional breakpoints
-// 
+//
 // Moved here from b-variables file so that we can load this partial
 // after USWDS variables and before other partials.
 //===========================
+// @media single-keyword helpers
+$small: new-breakpoint(min-width $small-screen $grid-columns-small);
+$medium: new-breakpoint(min-width $medium-screen $grid-columns-medium);
+$large: new-breakpoint(min-width $large-screen $grid-columns-large);
```
[the remark this answers is not in the document]
Comment (ncksllvn) [reply src/sass/base/_b-breakpoints.scss]: This actually shouldn't change our breakpoints at all. Those declaration should really just reset things so that USWDS can be imported. This reason why is sort of confusing...

So, Foundation defines those variables, but doesn't actually use them - they only define them because previous version of Foundation did use them, and they wanted to be backwards-compatible for apps that are dependent on them. 

Then, the new version of USWDS defined those variables too, and passes them as an argument to a mixin called `media`.  Everything should have been fine since we were importing USWDS after Foundation, except that USWDS was initializing the variables with that `!default` directive, so Foundation's variable declaration were overriding USWDS's. This meant that when USWDS tried using that `media` mixin, an error would be thrown about a bad parameter. That's why we define them there - it's the same code as USWDS, but that that `!default` flag.

So basically it goes like this -

1. Import Foundation 
  - -> defines small, medium, large
2. Import USWDS variables 
  - -> defines small, medium, and large with the `!default` directive, so Foundation's override them.
3. Import our custom breakpoints file, which redeclares USWDS's small, medium, and large without the `!default` flag
4. Import the rest of USWDS, which will now work as expected.

File: config/webpack.config.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (jbalboni) [reply config/webpack.config.js]: As far as I know, these don't really do anything right now. They allow us to import directly from bourbon/neat, I think, but we don't do that.

[the remark this answers is not in the document]
Comment (ncksllvn) [reply config/webpack.config.js]: Sweet! So maybe we can leave it commented out for now, and install them later if we decide that we need them as direct imports instead of from USWDS's install folder.

[the remark this answers is not in the document]
Comment (cehsu) [reply config/webpack.config.js]: Thanks @ncksllvn @jbalboni.

File: src/sass/base/_b-breakpoints.scss (hunk 3 above)
[the remark this answers is not in the document]
Comment (cehsu) [reply src/sass/base/_b-breakpoints.scss]: Thanks for the detailed explanation @ncksllvn.

**Review by phiden — APPROVED** [review]

**Comment by ncksllvn** [comment]
This passes our e2e tests as well visual regression testing, wahoo 🎉 

```
...
 ✔ Passed [ok]: http://localhost:3001/disability-benefits/conditions/exposure-to-hazardous-materials/agent-orange/diseases/
 ✔ Passed [ok]: http://localhost:3001/disability-benefits/conditions/exposure-to-hazardous-materials/agent-orange/navy-coast-guard/
 ✔ Passed [ok]: http://localhost:3001/disability-benefits/conditions/exposure-to-hazardous-materials/agent-orange/non-hodgkins/
 ✔ Passed [ok]: http://localhost:3001/disability-benefits/conditions/exposure-to-hazardous-materials/agent-orange/registry-health-exam/
 ✔ Passed [ok]: http://localhost:3001/disability-benefits/conditions/exposure-to-hazardous-materials/agent-orange/service-inside/
 ✔ Passed [ok]: http://localhost:3001/disability-benefits/conditions/exposure-to-hazardous-materials/agent-orange/service-outside/
 ✔ Passed [ok]: http://localhost:3001/disability-benefits/conditions/exposure-to-hazardous-materials/agent-orange/test-storage/
 ✔ Passed [ok]: http://localhost:3001/disability-benefits/conditions/exposure-to-hazardous-materials/agent-orange/thailand-military-bases/
 ✔ Passed [ok]: http://localhost:3001/disability-benefits/conditions/exposure-to-hazardous-materials/agent-orange/water-vietnam/

OK. 253 assertions passed. (7m 3s / 423821ms)
```

File: src/js/edu-benefits/education-wizard.js [hunk 4]
```diff
@@ -2,7 +2,7 @@ function toggleClass(element, className) {
   element.classList.toggle(className);
 }
 function openState(element) {
-  element.setAttribute('hidden', false);
+  element.removeAttribute('hidden');
```
Comment (jbalboni) [inline src/js/edu-benefits/education-wizard.js]: This file has been deleted in master, so you may get a conflict.

**Review by jbalboni — APPROVED** [review]

File: src/js/edu-benefits/education-wizard.js (hunk 4 above)
Comment (ncksllvn) [reply src/js/edu-benefits/education-wizard.js]: Ah gotcha, I heard that mentioned. I'm gonna revert this then and do an ugly override on `[hidden=false` instead. That's probably better anyway.

**Review by cehsu — APPROVED** [review]

**Comment by jbalboni** [comment]
Alerts have a border and some buttons have more rounded corners:

Staging:
screen shot 2017-10-26 at 2 04 27 pm
Review instance
screen shot 2017-10-26 at 2 03 00 pm

File: src/sass/_shame.scss [hunk 5]
```diff
@@ -201,12 +201,30 @@ body {
...
+  position: absolute;
+  left: auto;
+}
+
+// USWDS uses this:
+// [hidden] { display: none !important; }
+// However, some apps set the property to false and expect the element to be visible.
+// So, we do this shameful override to get around it.
+[hidden=false]{
+  display: block !important;
```
Comment (U-DON) [inline src/sass/_shame.scss]: To avoid shameful overrides, will this be resolved in a later PR where you do `removeAttribute` instead of setting `hidden` to `false`?

**Review by U-DON — APPROVED** [review]
LGTM. Will have to go through the Kudos products later to compare the visual changes.

**Comment by jbalboni** [comment]
We may want to fix the sizing of the alert icon here:

screen shot 2017-10-26 at 2 21 57 pm

**Comment by jbalboni** [comment]
This is broken on claim status:

http://c49169e4b77047b52cfd83831bd0e1d9.review.vetsgov-internal/track-claims/your-claims/600089290/files

screen shot 2017-10-26 at 2 24 12 pm

File: src/sass/_shame.scss (hunk 5 above)
Comment (ncksllvn) [reply src/sass/_shame.scss]: That was my first instinct (there's a revert commit on here where I was doing that previously). The file that toggles the `hidden` property is being removed though, and I don't know what it's doing instead, so I figured this was the safer route. But depending on what that file looks like we should be able to remove this rule.

**Comment by jbalboni** [comment]
Form review pages don't open for some reason:

screen shot 2017-10-26 at 2 31 47 pm

(This works on staging)

**Comment by jbalboni** [comment]
Button max width is being set again:

screen shot 2017-10-26 at 2 32 33 pm

**Comment by U-DON** [comment]
Does the visual regression testing not capture those differences?

**Comment by ncksllvn** [comment]
@jbalboni great job finding these!

@U-DON as it is now, it doesn't scroll the page, but that's the next thing I'm going to add. I wanted to get this foundation down pat before adding on.

**Comment by jbalboni** [comment]
@ncksllvn I think I'm done with Rainbows products (forms and claim status). If you want help fixing any of those, let me know.

**Comment by ncksllvn** [comment]
@jbalboni any help is definitely appreciated. I updated the description with more todo items if there's any you want to take

**Comment by ncksllvn** [comment]
Any bugs can be added as todo items here:

department-of-veterans-affairs/vets.gov-team/issues/5622

**Comment by phiden** [comment]
Accidentally merged while trying to do something else; sorry about that.
